server: fix a rare deadlock during tenant server shutdown (deflakes TestServerStartStop)

knz opened the pull request:
Two patches in this PR.

----
Fixes #100989.

Prior to this PR, there was a rare race condition when a hard shutdown of the KV node (via `stopper.Stop`) would catch up with a graceful shutdown of a secondary tenant server (e.g. via `ALTER TENANT STOP SERVICE`) and cause a deadlock.

This would likely never happen on a running server, because there a node shutdown/restart would wait for the graceful drain to complete, and in that case there is no deadlock. The issue is more prevalent in tests however and may cause the tests to time out.

This PR fixes the root cause of the issue. The actual fix is in the 8th commit; all the commits before that are cosmetic or helps to debugging.

In addition, the fix also revealed a bug in the handling of memory monitors upon ungraceful termination. This will be tracked in #101297.

----
Fixes #101403.

Prior to this patch, the task that was responsible for propagating a
graceful drain captured a variable supposed to reference the tenant
server, but before this variable was assigned.

This created a possible race condition, in the unlikely case case
where the server startup would fail _and_ a graceful drain would be
requested, concurrently.

This patch fixes it by only starting to propagate graceful drains
after the server is fully initialized. (But before it starts accepting
clients, so that we don't create a window of time where clients can
connects but graceful drains don't propagate.)

This is also achieved by extracting the two shutdown tasks into
separate functions, to clarify the flow of parameters.

knz committed:
server: lift the orchestration code under an interface

This commit extracts the core logic from
`server_controller_orchestration.go` into an
interface (`serverOrchestrator`) that can be mocked. We will use this
in testing.

This extraction is also useful because it exposes a few shortcomings
in the current implementation (wrt server shutdown). This makes it
easier to fix them in a later commit.

Release note: None

knz committed:
server: move the main tenant orchestration logic to its own file

This moves `(o *channelOrchestrator) startControlledServer` alongside
the other methods of `channelOrchestrator`.
There is further no code change.

Release note: None

knz committed:
server: split graceful vs ungraceful stop for tenant servers

Prior to this patch, immediate shutdown requests for secondary tenant
servers were serviced using the graceful path. This was incorrect, and
resulted in excessively long shutdown sequences in tests in some
cases.

This patch fixes it.

Release note: None

knz committed:
server: extra logging for server controller

Release note: None

knz committed:
server: avoid naked writes to a channel

The sever controller was writing to
`useGracefulDrainDuringTenantShutdown` unconditionally. As a result,
if two or more expedited stops caught up with a graceful stop, some of
the writes to the channel could be blocked. This would never happen on
a running system, but could be exercised in tests.

This patch fixes this by making the write unblocking. This generally
makes the code more readable anyway, since naked channel writes are
bound to raise eyebrows anyways.

Release note: None

knz committed:
server: add some shutdown logging

This helps while troubleshooting tests.

Release note: None

knz committed:
server: fix a possible deadlock during tenant server shutdown

Prior to this patch, the server orchestrator was unable to notice when
the surrounding server was stopping after it had started a graceful
drain already. This is because once the `propagate-close` task (and
its `select`) monitoring various stop signals noticed a request for
graceful drains, it would stop monitoring the other stop signals to
focus exclusively on the graceful drain; thereby missing when the
stopper for the surrounding server was quiescing.

The fix is to run the monitor for graceful and ungraceful shutdowns in
separate tasks, which this patch does.

This goes 80% of the way towards de-flaking `TestServerStartStop` in
`pkg/ccl/serverccl`, which was exhibiting the deadlock described
above. With this fix in place, the deadlock disappears entirely.
However, this reveals _another_ bug in the SQL layer which also needs
to be fixed to consider `TestServerStartStop` stable. We will do this
in a separate commit.

Release note: None

knz committed:
serverccl: add additional logging to `TestServerStartStop`

This also ensures the test fails quickly if there is a deadlock while
the server is shutting down.

(This makes the timeout for this test shorter than the standard
timeout for the stopper.Stop method, which is 15 minutes.)

Release note: None

knz committed:
util/mon: better logging

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101296)
<!-- Reviewable:end -->

knz committed:
sqlstats: paper over a memory monitor bug

When a hard stop (e.g. test exists) catches up with an ongoing
graceful drain, some bytes do not get released properly in a memory
monitor. This triggers a panic during shutdown, since the byte monitor
code verifies that all allocated bytes have been released.

This bug is relatively hard to trigger because in most cases,
a server is shut down either only using a graceful drain, or only
using a hard stop, but not both.

`TestServerStartStop` happens to do this and this is where this
problem was caught. We are now tracking it as issue #101297.

Until that issue is fixed, this commit papers over the problem by
removing the assertion in the byte monitor.

Release note: None

knz committed:
server: avoid a race condition in tenant server orchestration

Prior to this patch, the task that was responsible for propagating a
graceful drain captured a variable supposed to reference the tenant
server, but before this variable was assigned.

This created a possible race condition, in the unlikely case case
where the server startup would fail _and_ a graceful drain would be
requested, concurrently.

This patch fixes it by only starting to propagate graceful drains
after the server is fully initialized. (But before it starts accepting
clients, so that we don't create a window of time where clients can
connects but graceful drains don't propagate.)

This is also achieved by extracting the two shutdown tasks into
separate functions, to clarify the flow of parameters.

Release note: None

jeffswenson reviewed (approved):
LGTM

knz commented:
TFYR!

bors r=JeffSwenson
